### DISCUSSION TITLE
Angular and web components

### DISCUSSION THREAD

**Discussion by sijakret** [lit/lit#2242] [Ideas]
Hi Everyone,

first of all - I am a big fan of the lit ecosystem. It is by far my favorite way to create components and applications for the web platform!

I am opening this thread because I am a bit desperate about the following issue:

In our organization we created a lit-based web component library. The main driver to use lit was to make the library translate across different tech stacks.
Now, it turns out that most projects in the organization are built using Angular. Unfortunately angular is kind of a hostile environment for web components.

If you want to use web components in angular, you need to enable a `CUSTOM_ELEMENTS_SCHEMA` that essentially opts out of all type checks for components with dash-style-names (including angular components). This is a big drawback during development since you lose quite some static checks there.

There are really no technical reason for this but its mainly due to the fact that the angular compiler has a hardcoded (kid you not) dictionary of known html tags and their props that is not extensible. I have been [trying to keep a discussion alive and also sketched out solutions to fix it in angular](https://github.com/angular/angular/issues/12045) with limited success.

In my organization people are now resorting to writing _angular wrappers_ that host the web components which somehow feels really wrong. It adds an additional indirection, is just one more thing to maintain and kind of defeats the purpose of web components.

The point I am trying to get to is:

My understanding is that both lit and angular are technologies driven by google. My understanding is also that web components are seen as kind of a future-proof way to create reusable modules on the web. The reality is, however that many enterprise applications use Angular and currently it is really impractical to use a web component library in angular.
Web components are pretty much a prefect fit for design systems for organizations but as long as angular does not provide a viable path to use them there will always be people arguing against them for that very reason.
You cannot imagine the amount of discussions we are having on this topic. If it wasn't for a handful of people my org would have dropped web components already.

So, since everyone here works at google, is there anything you guys can do to help push this integration? It would get a lot of discussions out of the way and would likely also give web component adoption a boost. :)

What are your thoughts?
How can we generate some traction here?

**Comment by damsalem** [comment]
Someone sent me to read this discussion, it seems valid according to the [Angular docs](https://angular.io/api/core/CUSTOM_ELEMENTS_SCHEMA).

After some digging, I'm thinking this implementation should solve that [https://manuel-rauber.com/2021/02/23/youre-using-custom_elements_schema-wrong/](https://manuel-rauber.com/2021/02/23/youre-using-custom_elements_schema-wrong/)

**Reply by sijakret** [reply]
This is a workaround and not a real solution. To write a wrapper you need to
* redeclare the component API in angular
* accept an additional layer in angular that does unnecessary stuff at runtime 
* deal with angular which should not be necessary in the first place when publishing a design system built as web components

Of course this approach can be handled to a large extent using code generation. But why? It is pure overhead. This issue should be solved on the angular side and not on the WC side! I have made several arguments in the thread over at angular on why I think so and why I think it should not be hard.

**Comment by drk-mtr** [comment]
I would also like to understand how to get proper type checking, such that the compiler will tell me if I try to set a value for a property that does not exist on the web component. I find it rather bizarre that there's a way to generate custom elements from Angular components, but seemingly no way to import them that results in a nice developer experience.

One thing I've just found out: if the `tsconfig.json` for the Angular project has `"moduleResolution": "node"` under `compilerOptions` then I was not getting intellisense when importing web components from the web component library. When I set this to `"moduleResolution": "bundler"`, things start working as expected. 

I'm still not sure exactly why this is the case, and whether this would be more appropriate in the Angular docs or in some Lit "integrating with other frameworks" guide.

**Comment by withinsight** [comment]
I similarly have product teams saying they'll be missing things like dependency injection and directives when adopting Web Components. Its mind-boggling the Angular team hasn't worked on an interface to allow custom Angular things to work with Web Components built in Lit. Custom Elements was kind of the reverse (i.e. start with Angular and convert them to WC's, but you have to take the larger footprint of Angular in this scenario. The reverse, enabling Lit components to be used in an Angular app, and engineers being able to use their existing directives or dependencies seems like it would be helpful. Although, I'm not sure if its just overkill.

**Comment by irfanullahjan** [comment]
The wrapper approach doesn't sound right to me. In my opinion, the web components should be consumed in the frameworks directly. The type-safety should be the job of Angular/Vue/React compiler.

With JSX, the situation is a bit better. Angular doesn't work with JSX unfortunately. So in Angular templates, things like autocompletion should be provided by the Angular Language service extension, in VSCode for example.

**Comment by geoff-ponnath-44115** [comment]
I added a [feature request](https://github.com/angular/angular/issues/63015#issuecomment-3167847402) to make angular more friendly to different types of FACEs (Form associated custom elements). Currently the `ngDefaultControl` directive will allow a text input FACE to work with angular forms, but _there is nothing for checkboxes or other control types other than writing custom directives_ like this global one that targets `<my-checkbox>` elements using `ngModel`.

```
@Directive({
  selector: "my-checkbox[ngModel]",
  providers: [{
    provide: NG_VALUE_ACCESSOR,
    useExisting: forwardRef(() => MyCheckboxDirective),
    multi: true
  }],
})
export class MyCheckboxDirective extends CheckboxControlValueAccessor { }
```
Its not the ideal situation but would help a lot in that regard and reduce the need for wrappers.

**Comment by pfteter** [comment]
Any updates on this:
https://www.youtube.com/watch?v=CMZHUPSmZx4&t=624s

**Reply by justinfagnani** [reply]
@pfteter that might be a better question for somewhere on the Angular project.

**Reply by leblancmeneses** [reply]
with signals + hydration +  zoneless stable will we get a version of:

```
// https://lit.dev/docs/frameworks/react/
import {createComponent} from '@lit/angular';
```
?

Would love a version that provided Declarative Shadow DOM with angular elements + hydration delivered over Analog.js.
